File: ios/runanywhere-sdks-latest/examples/ios/RunAnywhereAI/RunAnywhereAI/Features/Voice/SpeechToTextView.swift
import SwiftUI
import RunAnywhere
#if os(macOS)
import AppKit
#endif

/// Dedicated Speech-to-Text view with real-time transcription
/// This view is purely focused on UI - all business logic is in STTViewModel
struct SpeechToTextView: View {
    @StateObject private var viewModel = STTViewModel()
    @State private var showModelPicker = false
    @State private var breathingAnimation = false

    private var hasModelSelected: Bool {
        viewModel.selectedModelName != nil
    }

    private var statusMessage: String {
        ""
    }

    private var waveHeights: [CGFloat] {
        breathingAnimation
            ? [24, 40, 32, 48, 28]
            : [16, 24, 20, 28, 18]
    }

    var body: some View {
        Group {
            NavigationView {
                ZStack {
                    VStack(spacing: 0) {
                        // Mode selection - Modern pill button style
                        if hasModelSelected {
                        HStack(spacing: 8) {
                            modeButton(.batch, title: "Batch", subtitle: "Record")
                            modeButton(.live, title: "Live", subtitle: "Stream")
                            modeButton(.hybrid, title: "Hybrid", subtitle: "Cloud")
                        }
                        .padding(.horizontal, 16)
                        .padding(.top, 12)
                        .padding(.bottom, 8)

                        if viewModel.selectedMode == .hybrid {
                            hybridConfigurationSection
                                .padding(.horizontal, 16)
                                .padding(.bottom, 8)
                        }
                        }

                        // Main content - only enabled when model is selected
                        if hasModelSelected {
                        if viewModel.transcription.isEmpty && !viewModel.isRecording && !viewModel.isTranscribing {
                            // Ready state - Modern minimal design
                            VStack(spacing: 0) {
                                Spacer()

                                VStack(spacing: 48) {
                                    // Minimal waveform visualization
                                    HStack(spacing: 4) {
                                        ForEach(0..<5) { index in
                                            RoundedRectangle(cornerRadius: 8)
                                                .fill(
                                                    LinearGradient(
                                                        colors: [
                                                            AppColors.primaryAccent.opacity(0.8),
                                                            AppColors.primaryAccent.opacity(0.4)
                                                        ],
                                                        startPoint: .top,
                                                        endPoint: .bottom
                                                    )
                                                )
                                                .frame(width: 6, height: waveHeights[index])
                                                .animation(
                                                    .easeInOut(duration: 0.8)
                                                        .repeatForever(autoreverses: true)
                                                        .delay(Double(index) * 0.1),
                                                    value: breathingAnimation
                                                )
                                        }
                                    }

                                    // Clean typography
                                    VStack(spacing: 12) {
                                        Text("Ready to transcribe")
                                            .font(.system(size: 24, weight: .semibold, design: .rounded))
                                            .foregroundColor(.primary)

                                        Text(readyModeDescription)
                                            .font(.system(size: 15, weight: .regular))
                                            .foregroundColor(.secondary)
                                    }
                                }

                                Spacer()
                            }
                            .onAppear {
                                breathingAnimation = true
                            }
                        } else if viewModel.isTranscribing && viewModel.transcription.isEmpty {
                            // Processing state - Clean and centered
                            VStack(spacing: 0) {
                                Spacer()

                                ProgressView()
                                    .scaleEffect(1.2)
                                    .padding(.bottom, 12)

                                Text("Transcribing...")
                                    .font(.subheadline)
                                    .foregroundColor(.secondary)

                                Spacer()
                            }
                        } else {
                            // Transcription display with ScrollView
                            ScrollView {
                                VStack(alignment: .leading, spacing: 12) {
                                        HStack {
                                            Text("Transcription")
                                                .font(.headline)
                                                .foregroundColor(.primary)

                                            Spacer()

                                            if viewModel.isRecording {
                                                HStack(spacing: 6) {
                                                    Circle()
                                                        .fill(AppColors.statusRed)
                                                        .frame(width: 8, height: 8)
                                                    Text("RECORDING")
                                                        .font(.caption2)
                                                        .fontWeight(.bold)
                                                        .foregroundColor(AppColors.statusRed)
                                                }
                                                .padding(.horizontal, 8)
                                                .padding(.vertical, 4)
                                                .background(AppColors.statusRed.opacity(0.1))
                                                .cornerRadius(AppSpacing.cornerRadiusSmall)
                                            } else if viewModel.isTranscribing {
                                                HStack(spacing: 6) {
                                                    ProgressView()
                                                        .scaleEffect(0.6)
                                                    Text("TRANSCRIBING")
                                                        .font(.caption2)
                                                        .fontWeight(.bold)
                                                        .foregroundColor(AppColors.statusOrange)
                                                }
                                                .padding(.horizontal, 8)
                                                .padding(.vertical, 4)
                                                .background(AppColors.statusOrange.opacity(0.1))
                                                .cornerRadius(AppSpacing.cornerRadiusSmall)
                                            }
                                        }

                                        Text(viewModel.transcription)
                                            .font(.body)
                                            .foregroundColor(.primary)
                                            .padding()
                                            .frame(maxWidth: .infinity, alignment: .leading)
                                            #if os(iOS)
                                            .background(Color(.secondarySystemBackground))
                                            #else
                                            .background(Color(NSColor.controlBackgroundColor))
                                            #endif
                                            .cornerRadius(AppSpacing.cornerRadiusXLarge)

                                        if let routing = viewModel.hybridRouting {
                                            hybridRoutingSummary(routing)
                                        }
                                }
                                .padding()
                            }
                        }

                        // Controls
                        VStack(spacing: 16) {
                            // Error message
                            if let error = viewModel.errorMessage {
                                Text(error)
                                    .font(.caption)
                                    .foregroundColor(AppColors.statusRed)
                                    .multilineTextAlignment(.center)
                                    .padding(.horizontal)
                            }

                            // Audio level indicator
                            if viewModel.isRecording {
                                AdaptiveAudioLevelIndicator(level: viewModel.audioLevel)
                            }

                            // Record button
                            AdaptiveMicButton(
                                isActive: viewModel.isRecording,
                                isPulsing: false,
                                isLoading: viewModel.isProcessing || viewModel.isTranscribing,
                                activeColor: AppColors.statusRed,
                                inactiveColor: viewModel.isTranscribing
                                    ? AppColors.statusOrange
                                    : AppColors.primaryAccent,
                                icon: viewModel.isRecording ? "stop.fill" : "mic.fill"
                            ) {
                                Task {
                                    await viewModel.toggleRecording()
                                }
                            }
                            .disabled(
                                viewModel.selectedModelName == nil ||
                                viewModel.isProcessing ||
                                viewModel.isTranscribing
                            )
                            .opacity(
                                viewModel.selectedModelName == nil ||
                                viewModel.isProcessing ||
                                viewModel.isTranscribing ? 0.6 : 1.0
                            )

                            if !statusMessage.isEmpty {
                                Text(statusMessage)
                                    .font(.caption)
                                    .foregroundColor(.secondary)
                            }
                        }
                        .padding()
                        #if os(iOS)
                        .background(Color(.systemBackground))
                        #else
                        .background(Color(NSColor.windowBackgroundColor))
                        #endif
                        } else {
                            // No model selected - show onboarding
                            Spacer()
                        }
                    }

                    // Overlay when no model is selected
                if !hasModelSelected && !viewModel.isProcessing {
                    ModelRequiredOverlay(
                        modality: .stt
                    ) { showModelPicker = true }
                }
                }
            .navigationTitle(hasModelSelected ? "Speech to Text" : "")
            #if os(iOS)
            .navigationBarTitleDisplayModeCompat(.inline)
            .navigationBarHidden(!hasModelSelected)
            #endif
            .toolbar {
                if hasModelSelected {
                    #if os(iOS)
                    ToolbarItem(placement: .navigationBarTrailing) {
                        modelButton
                    }
                    #else
                    ToolbarItem(placement: .automatic) {
                        modelButton
                    }
                    #endif
                }
            }
            }
        #if os(iOS)
        .navigationViewStyle(.stack)
        #endif
        .adaptiveSheet(isPresented: $showModelPicker) {
            ModelSelectionSheet(context: .stt) { model in
                Task {
                    await viewModel.loadModelFromSelection(model)
                }
            }
        }
        .onAppear {
            Task {
                await viewModel.initialize()
            }
        }
        .onDisappear {
            viewModel.cleanup()
        }
        }
    }

    // MARK: - View Components

    private var readyModeDescription: String {
        switch viewModel.selectedMode {
        case .batch:
            return "Record first, then transcribe"
        case .live:
            return "Real-time transcription"
        case .hybrid:
            return "On-device first with cloud fallback"
        }
    }

    private func modeButton(_ mode: STTMode, title: String, subtitle: String) -> some View {
        Button {
            withAnimation(.spring(response: 0.3, dampingFraction: 0.7)) {
                viewModel.selectedMode = mode
            }
        } label: {
            VStack(spacing: 4) {
                Text(title)
                    .font(.system(size: 13, weight: .medium))
                Text(subtitle)
                    .font(AppTypography.system10)
                    .opacity(0.7)
            }
            .frame(maxWidth: .infinity)
            .padding(.vertical, 12)
            .background(
                viewModel.selectedMode == mode
                    ? AppColors.primaryAccent.opacity(0.15)
                    : Color.clear
            )
            .foregroundColor(
                viewModel.selectedMode == mode
                    ? AppColors.primaryAccent
                    : .secondary
            )
            .cornerRadius(AppSpacing.cornerRadiusXLarge)
            .overlay(
                RoundedRectangle(cornerRadius: 12)
                    .stroke(
                        viewModel.selectedMode == mode
                            ? AppColors.primaryAccent.opacity(0.3)
                            : AppColors.statusGray.opacity(0.2),
                        lineWidth: 1
                    )
            )
        }
    }

    private var hybridConfigurationSection: some View {
        VStack(alignment: .leading, spacing: 10) {
            HStack(spacing: 8) {
                TextField("provider", text: $viewModel.cloudProvider)
                TextField("model", text: $viewModel.cloudModel)
            }
            TextField("cloud registry id", text: $viewModel.cloudProviderId)
            SecureField("cloud API key", text: $viewModel.cloudAPIKey)
            TextField("language", text: $viewModel.cloudLanguageCode)
            Toggle("Prefer online", isOn: $viewModel.hybridPreferOnline)
            Toggle("Require network", isOn: $viewModel.hybridRequireNetwork)
            VStack(alignment: .leading, spacing: 4) {
                Text("Fallback threshold \(viewModel.hybridConfidenceThreshold, specifier: "%.2f")")
                    .font(.caption)
                    .foregroundColor(.secondary)
                Slider(value: $viewModel.hybridConfidenceThreshold, in: 0.0...1.0, step: 0.05)
            }
        }
        .textFieldStyle(.roundedBorder)
        .font(.caption)
        .padding(12)
        #if os(iOS)
        .background(Color(.secondarySystemBackground))
        #else
        .background(Color(NSColor.controlBackgroundColor))
        #endif
        .cornerRadius(AppSpacing.cornerRadiusXLarge)
    }

    private func hybridRoutingSummary(_ routing: HybridRoutedMetadata) -> some View {
        VStack(alignment: .leading, spacing: 6) {
            Text("Routing")
                .font(.subheadline)
                .fontWeight(.semibold)
            Text("Chosen: \(routing.chosenModelID.isEmpty ? "unknown" : routing.chosenModelID)")
            Text("Fallback: \(routing.wasFallback ? "yes" : "no")")
            if !routing.primaryErrorMessage.isEmpty {
                Text("Primary: \(routing.primaryErrorMessage)")
            }
        }
        .font(.caption)
        .foregroundColor(.secondary)
        .padding(12)
        #if os(iOS)
        .background(Color(.tertiarySystemBackground))
        #else
        .background(Color(NSColor.controlBackgroundColor).opacity(0.8))
        #endif
        .cornerRadius(AppSpacing.cornerRadiusLarge)
    }

    private var modelButton: some View {
        Button {
            showModelPicker = true
        } label: {
            HStack(spacing: 6) {
                // Model logo instead of cube icon
                if let modelName = viewModel.selectedModelName {
                    Image(getModelLogo(for: modelName))
                        .resizable()
                        .aspectRatio(contentMode: .fit)
                        .frame(width: 36, height: 36)
                        .cornerRadius(AppSpacing.cornerRadiusSmall)
                } else {
                    Image(systemName: "cube")
                        .font(AppTypography.system14)
                }

                if let modelName = viewModel.selectedModelName {
                    VStack(alignment: .leading, spacing: 2) {
                        Text(modelName.shortModelName())
                            .font(.caption)
                            .fontWeight(.medium)
                            .lineLimit(1)

                        // Framework indicator
                        if let framework = viewModel.selectedFramework {
                            HStack(spacing: 3) {
                                Image(systemName: frameworkIcon(for: framework))
                                    .font(.system(size: 7))
                                Text(framework.displayName)
                                    .font(.system(size: 8, weight: .medium))
                            }
                            .foregroundColor(frameworkColor(for: framework))
                        }
                    }
                } else {
                    Text("Select Model")
                        .font(.caption)
                }
            }
        }
    }


    private func frameworkIcon(for framework: InferenceFramework) -> String {
        switch framework {
        case .onnx: return "square.stack.3d.up"
        case .foundationModels: return "apple.logo"
        default: return "cube"
        }
    }

    private func frameworkColor(for framework: InferenceFramework) -> Color {
        switch framework {
        case .onnx: return AppColors.primaryPurple
        case .foundationModels: return .primary
        default: return AppColors.statusGray
        }
    }
}

File: ios/runanywhere-sdks-latest/examples/flutter/RunAnywhereAI/lib/features/voice/speech_to_text_view.dart
import 'dart:async';

import 'package:flutter/material.dart';
import 'package:flutter/services.dart';
import 'package:runanywhere/runanywhere.dart' as sdk;
import 'package:runanywhere_ai/core/design_system/app_colors.dart';
import 'package:runanywhere_ai/core/design_system/app_spacing.dart';
import 'package:runanywhere_ai/core/design_system/typography.dart';
import 'package:runanywhere_ai/core/services/permission_service.dart';
import 'package:runanywhere_ai/features/models/model_selection_sheet.dart';
import 'package:runanywhere_ai/features/models/model_status_components.dart';
import 'package:runanywhere_ai/features/models/model_types.dart';
import 'package:runanywhere_ai/features/voice/stt_view_model.dart';

/// SpeechToTextView (mirroring iOS SpeechToTextView.swift)
///
/// Dedicated STT view with batch/live/hybrid mode support and real-time
/// transcription. Purely UI — all business logic lives in [STTViewModel].
class SpeechToTextView extends StatefulWidget {
  const SpeechToTextView({super.key});

  @override
  State<SpeechToTextView> createState() => _SpeechToTextViewState();
}

class _SpeechToTextViewState extends State<SpeechToTextView> {
  final STTViewModel _viewModel = STTViewModel();

  // Cloud-config text fields are view-owned controllers seeded from the
  // ViewModel defaults; edits are pushed back via onChanged.
  late final TextEditingController _cloudProviderController;
  late final TextEditingController _cloudModelController;
  late final TextEditingController _cloudProviderIdController;
  late final TextEditingController _cloudApiKeyController;
  late final TextEditingController _cloudLanguageController;

  @override
  void initState() {
    super.initState();
    _cloudProviderController =
        TextEditingController(text: _viewModel.cloudProvider);
    _cloudModelController = TextEditingController(text: _viewModel.cloudModel);
    _cloudProviderIdController =
        TextEditingController(text: _viewModel.cloudProviderId);
    _cloudApiKeyController = TextEditingController(text: _viewModel.cloudApiKey);
    _cloudLanguageController =
        TextEditingController(text: _viewModel.cloudLanguageCode);
    unawaited(_viewModel.initialize());
  }

  @override
  void dispose() {
    _cloudProviderController.dispose();
    _cloudModelController.dispose();
    _cloudProviderIdController.dispose();
    _cloudApiKeyController.dispose();
    _cloudLanguageController.dispose();
    _viewModel.dispose();
    super.dispose();
  }

  void _showModelSelectionSheet() {
    unawaited(showModalBottomSheet<void>(
      context: context,
      isScrollControlled: true,
      useSafeArea: true,
      builder: (sheetContext) => ModelSelectionSheet(
        context: ModelSelectionContext.stt,
        onModelSelected: (model) async {
          await _viewModel.loadModelFromSelection(model);
          // Surface load failures via SnackBar so the user sees them even if
          // the inline error text is below the fold or behind a sheet
          // animation.
          final error = _viewModel.errorMessage;
          if (error != null && mounted) {
            ScaffoldMessenger.of(context).showSnackBar(
              SnackBar(
                content: Text('STT load failed: $error'),
                duration: const Duration(seconds: 6),
              ),
            );
          }
        },
      ),
    ));
  }

  Future<void> _toggleRecording() async {
    if (!_viewModel.isRecording) {
      // Request STT permissions (microphone + speech recognition on iOS)
      // before starting — permission UI needs a BuildContext, so it stays in
      // the view.
      final hasPermission =
          await PermissionService.shared.requestSTTPermissions(context);
      if (!hasPermission) {
        _viewModel
            .reportError('Microphone permission required for recording');
        return;
      }
    }
    await _viewModel.toggleRecording();
  }

  void _copyToClipboard() {
    if (_viewModel.transcription.isEmpty) return;
    unawaited(
        Clipboard.setData(ClipboardData(text: _viewModel.transcription)));
    ScaffoldMessenger.of(context).showSnackBar(
      const SnackBar(content: Text('Copied to clipboard')),
    );
  }

  @override
  Widget build(BuildContext context) {
    return ListenableBuilder(
      listenable: _viewModel,
      builder: (context, _) {
        final hasModelSelected = _viewModel.hasModelSelected;
        return Scaffold(
          appBar: AppBar(
            title: const Text('Speech to Text'),
            actions: [
              if (_viewModel.transcription.isNotEmpty)
                IconButton(
                  icon: const Icon(Icons.copy),
                  onPressed: _copyToClipboard,
                  tooltip: 'Copy',
                ),
              if (_viewModel.transcription.isNotEmpty)
                IconButton(
                  icon: const Icon(Icons.delete_outline),
                  onPressed: _viewModel.clearTranscription,
                  tooltip: 'Clear',
                ),
            ],
          ),
          body: Stack(
            children: [
              Column(
                children: [
                  // Model Status Banner
                  Padding(
                    padding: const EdgeInsets.all(AppSpacing.large),
                    child: ModelStatusBanner(
                      framework: _viewModel.selectedFramework,
                      modelName: _viewModel.selectedModelName,
                      isLoading: _viewModel.isProcessing && !hasModelSelected,
                      onSelectModel: _showModelSelectionSheet,
                    ),
                  ),

                  // Mode selector (only when model is selected)
                  if (hasModelSelected) ...[
                    _buildModeSelector(),
                    _buildModeDescription(),
                    if (_viewModel.selectedMode == STTMode.hybrid)
                      _buildHybridConfigurationSection(),
                  ],

                  const Divider(),

                  // Main content
                  if (hasModelSelected) ...[
                    Expanded(child: _buildTranscriptionArea()),
                    const Divider(),
                    _buildControlsArea(),
                  ] else
                    const Expanded(child: SizedBox()),
                ],
              ),

              // Model required overlay
              if (!hasModelSelected && !_viewModel.isProcessing)
                ModelRequiredOverlay(
                  modality: ModelSelectionContext.stt,
                  onSelectModel: _showModelSelectionSheet,
                ),
            ],
          ),
        );
      },
    );
  }

  Widget _buildModeSelector() {
    return Padding(
      padding: const EdgeInsets.symmetric(horizontal: AppSpacing.large),
      child: SegmentedButton<STTMode>(
        segments: [
          for (final mode in STTMode.values)
            ButtonSegment(
              value: mode,
              label: Text(mode.displayName),
              icon: Icon(mode.icon),
            ),
        ],
        selected: {_viewModel.selectedMode},
        onSelectionChanged: _viewModel.isRecording
            ? null
            : (Set<STTMode> selection) {
                _viewModel.selectedMode = selection.first;
              },
      ),
    );
  }

  Widget _buildModeDescription() {
    return Padding(
      padding: const EdgeInsets.all(AppSpacing.large),
      child: Row(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          Icon(
            _viewModel.selectedMode.icon,
            size: AppSpacing.iconSmall,
            color: AppColors.textSecondary(context),
          ),
          const SizedBox(width: AppSpacing.xSmall),
          Text(
            _viewModel.selectedMode.description,
            style: AppTypography.caption(context).copyWith(
              color: AppColors.textSecondary(context),
            ),
          ),
          if (!_viewModel.supportsLiveMode &&
              _viewModel.selectedMode == STTMode.live) ...[
            const SizedBox(width: AppSpacing.smallMedium),
            Text(
              '(will use batch)',
              style: AppTypography.caption(context).copyWith(
                color: AppColors.statusOrange,
              ),
            ),
          ],
        ],
      ),
    );
  }

  /// Cloud routing configuration form (mirrors the iOS
  /// `hybridConfigurationSection`).
  Widget _buildHybridConfigurationSection() {
    return Container(
      margin: const EdgeInsets.symmetric(horizontal: AppSpacing.large),
      padding: const EdgeInsets.all(AppSpacing.mediumLarge),
      decoration: BoxDecoration(
        color: AppColors.backgroundGray6(context),
        borderRadius: BorderRadius.circular(AppSpacing.cornerRadiusCard),
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Row(
            children: [
              Expanded(
                child: _cloudTextField(
                  controller: _cloudProviderController,
                  hint: 'provider',
                  onChanged: (v) => _viewModel.cloudProvider = v,
                ),
              ),
              const SizedBox(width: AppSpacing.smallMedium),
              Expanded(
                child: _cloudTextField(
                  controller: _cloudModelController,
                  hint: 'model',
                  onChanged: (v) => _viewModel.cloudModel = v,
                ),
              ),
            ],
          ),
          const SizedBox(height: AppSpacing.smallMedium),
          _cloudTextField(
            controller: _cloudProviderIdController,
            hint: 'cloud registry id',
            onChanged: (v) => _viewModel.cloudProviderId = v,
          ),
          const SizedBox(height: AppSpacing.smallMedium),
          _cloudTextField(
            controller: _cloudApiKeyController,
            hint: 'cloud API key',
            obscureText: true,
            onChanged: (v) => _viewModel.cloudApiKey = v,
          ),
          const SizedBox(height: AppSpacing.smallMedium),
          _cloudTextField(
            controller: _cloudLanguageController,
            hint: 'language',
            onChanged: (v) => _viewModel.cloudLanguageCode = v,
          ),
          SwitchListTile(
            contentPadding: EdgeInsets.zero,
            dense: true,
            title: Text('Prefer online', style: AppTypography.caption(context)),
            value: _viewModel.hybridPreferOnline,
            onChanged: (v) => _viewModel.hybridPreferOnline = v,
          ),
          SwitchListTile(
            contentPadding: EdgeInsets.zero,
            dense: true,
            title:
                Text('Require network', style: AppTypography.caption(context)),
            value: _viewModel.hybridRequireNetwork,
            onChanged: (v) => _viewModel.hybridRequireNetwork = v,
          ),
          Text(
            'Fallback threshold '
            '${_viewModel.hybridConfidenceThreshold.toStringAsFixed(2)}',
            style: AppTypography.caption(context).copyWith(
              color: AppColors.textSecondary(context),
            ),
          ),
          Slider(
            value: _viewModel.hybridConfidenceThreshold,
            min: 0.0,
            max: 1.0,
            divisions: 20,
            onChanged: (v) => _viewModel.hybridConfidenceThreshold = v,
          ),
        ],
      ),
    );
  }

  Widget _cloudTextField({
    required TextEditingController controller,
    required String hint,
    required ValueChanged<String> onChanged,
    bool obscureText = false,
  }) {
    return TextField(
      controller: controller,
      obscureText: obscureText,
      style: AppTypography.caption(context),
      decoration: InputDecoration(
        hintText: hint,
        isDense: true,
        border: const OutlineInputBorder(),
        contentPadding: const EdgeInsets.symmetric(
          horizontal: AppSpacing.smallMedium,
          vertical: AppSpacing.smallMedium,
        ),
      ),
      onChanged: onChanged,
    );
  }

  Widget _buildTranscriptionArea() {
    return SingleChildScrollView(
      padding: const EdgeInsets.all(AppSpacing.large),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          if (_viewModel.transcription.isEmpty &&
              _viewModel.partialText.isEmpty &&
              !_viewModel.isRecording &&
              !_viewModel.isTranscribing)
            _buildReadyState()
          else if (_viewModel.isTranscribing &&
              _viewModel.transcription.isEmpty)
            _buildProcessingState()
          else
            _buildTranscriptionContent(),
        ],
      ),
    );
  }

  Widget _buildReadyState() {
    return Center(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          const SizedBox(height: AppSpacing.xxxLarge),
          Icon(
            Icons.mic,
            size: 64,
            color: AppColors.statusGreen.withValues(alpha: 0.5),
          ),
          const SizedBox(height: AppSpacing.large),
          Text(
            'Ready to transcribe',
            style: AppTypography.headline(context),
          ),
          const SizedBox(height: AppSpacing.smallMedium),
          Text(
            'Tap the microphone button to start recording',
            style: AppTypography.subheadline(context).copyWith(
              color: AppColors.textSecondary(context),
            ),
            textAlign: TextAlign.center,
          ),
        ],
      ),
    );
  }

  Widget _buildProcessingState() {
    return Center(
      child: Column(
        mainAxisAlignment: MainAxisAlignment.center,
        children: [
          const SizedBox(height: AppSpacing.xxxLarge),
          const SizedBox(
            width: 48,
            height: 48,
            child: CircularProgressIndicator(strokeWidth: 3),
          ),
          const SizedBox(height: AppSpacing.large),
          Text(
            'Processing audio...',
            style: AppTypography.headline(context),
          ),
          const SizedBox(height: AppSpacing.smallMedium),
          Text(
            'Transcribing your recording',
            style: AppTypography.subheadline(context).copyWith(
              color: AppColors.textSecondary(context),
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildTranscriptionContent() {
    return Column(
      crossAxisAlignment: CrossAxisAlignment.start,
      children: [
        // Header with status badge
        Row(
          children: [
            Text(
              'Transcription',
              style: AppTypography.headline(context),
            ),
            const Spacer(),
            RecordingStatusBadge(
              isRecording: _viewModel.isRecording,
              isTranscribing: _viewModel.isTranscribing,
            ),
          ],
        ),
        const SizedBox(height: AppSpacing.mediumLarge),

        // Transcription text area
        Container(
          width: double.infinity,
          padding: const EdgeInsets.all(AppSpacing.large),
          decoration: BoxDecoration(
            color: AppColors.backgroundGray6(context),
            borderRadius: BorderRadius.circular(AppSpacing.cornerRadiusCard),
          ),
          child: Column(
            crossAxisAlignment: CrossAxisAlignment.start,
            children: [
              if (_viewModel.transcription.isNotEmpty)
                Text(
                  _viewModel.transcription,
                  style: AppTypography.body(context),
                ),
              if (_viewModel.partialText.isNotEmpty)
                Text(
                  _viewModel.partialText,
                  style: AppTypography.body(context).copyWith(
                    color: AppColors.textSecondary(context),
                    fontStyle: FontStyle.italic,
                  ),
                ),
              if (_viewModel.transcription.isEmpty &&
                  _viewModel.partialText.isEmpty)
                Text(
                  'Listening...',
                  style: AppTypography.body(context).copyWith(
                    color: AppColors.textSecondary(context),
                  ),
                ),
            ],
          ),
        ),

        // Hybrid routing badge (where the transcript came from)
        if (_viewModel.hybridRouting != null) ...[
          const SizedBox(height: AppSpacing.mediumLarge),
          _buildHybridRoutingSummary(_viewModel.hybridRouting!),
        ],
      ],
    );
  }

  /// Routing decision summary (mirrors the iOS `hybridRoutingSummary`).
  Widget _buildHybridRoutingSummary(sdk.HybridRoutedMetadata routing) {
    final caption = AppTypography.caption(context).copyWith(
      color: AppColors.textSecondary(context),
    );
    return Container(
      width: double.infinity,
      padding: const EdgeInsets.all(AppSpacing.mediumLarge),
      decoration: BoxDecoration(
        color: AppColors.backgroundGray5(context),
        borderRadius: BorderRadius.circular(AppSpacing.cornerRadiusRegular),
      ),
      child: Column(
        crossAxisAlignment: CrossAxisAlignment.start,
        children: [
          Text(
            'Routing',
            style: AppTypography.captionMedium(context),
          ),
          const SizedBox(height: AppSpacing.xSmall),
          Text(
            'Chosen: '
            '${routing.chosenModelId.isEmpty ? "unknown" : routing.chosenModelId}',
            style: caption,
          ),
          Text(
            'Fallback: ${routing.wasFallback ? "yes" : "no"}',
            style: caption,
          ),
          if (routing.primaryErrorMessage.isNotEmpty)
            Text(
              'Primary: ${routing.primaryErrorMessage}',
              style: caption,
            ),
        ],
      ),
    );
  }

  Widget _buildControlsArea() {
    return Container(
      padding: const EdgeInsets.all(AppSpacing.large),
      child: Column(
        children: [
          // Error message
          if (_viewModel.errorMessage != null)
            Padding(
              padding: const EdgeInsets.only(bottom: AppSpacing.large),
              child: Text(
                _viewModel.errorMessage!,
                style: AppTypography.caption(context).copyWith(
                  color: AppColors.primaryRed,
                ),
                textAlign: TextAlign.center,
              ),
            ),

          // Audio level indicator
          if (_viewModel.isRecording)
            Padding(
              padding: const EdgeInsets.only(bottom: AppSpacing.large),
              child: AudioLevelIndicator(level: _viewModel.audioLevel),
            ),

          // Record button
          _buildRecordButton(),

          const SizedBox(height: AppSpacing.mediumLarge),

          // Status text
          Text(
            _viewModel.isTranscribing
                ? 'Processing transcription...'
                : _viewModel.isRecording
                    ? 'Tap to stop recording'
                    : 'Tap to start recording',
            style: AppTypography.caption(context).copyWith(
              color: AppColors.textSecondary(context),
            ),
          ),
        ],
      ),
    );
  }

  Widget _buildRecordButton() {
    final Color buttonColor;
    if (_viewModel.isRecording) {
      buttonColor = AppColors.primaryRed;
    } else if (_viewModel.isTranscribing) {
      buttonColor = AppColors.primaryOrange;
    } else {
      buttonColor = AppColors.primaryBlue;
    }

    return GestureDetector(
      onTap: !_viewModel.isProcessing &&
              !_viewModel.isTranscribing &&
              _viewModel.hasModelSelected
          ? _toggleRecording
          : null,
      child: Container(
        width: 72,
        height: 72,
        decoration: BoxDecoration(
          shape: BoxShape.circle,
          color: _viewModel.hasModelSelected ? buttonColor : AppColors.statusGray,
          boxShadow: [
            BoxShadow(
              color: buttonColor.withValues(alpha: 0.3),
              blurRadius: AppSpacing.shadowXLarge,
              offset: const Offset(0, 4),
            ),
          ],
        ),
        child: _viewModel.isProcessing || _viewModel.isTranscribing
            ? const Padding(
                padding: EdgeInsets.all(AppSpacing.xLarge),
                child: CircularProgressIndicator(
                  color: Colors.white,
                  strokeWidth: 2,
                ),
              )
            : Icon(
                _viewModel.isRecording ? Icons.stop : Icons.mic,
                color: Colors.white,
                size: 32,
              ),
      ),
    );
  }
}
